scrapy_webarchive/wacz/storages.py

```diff
@@ -0,0 +1,232 @@
+    AWS_VERIFY = None
+
+    def __init__(self, uri: str):
+        if not is_botocore_available():
+            raise NotConfigured("missing botocore library")
+
+        parse_result = urlparse(uri, allow_fragments=False)
+        if parse_result.scheme != "s3":
+            raise ValueError(f"Incorrect URI scheme in {uri}, expected 's3'")
+
+        self.bucket = parse_result.netloc
+        self.path = parse_result.path.lstrip("/")
+        self.s3_client = self._initialize_s3_client()
+        self.zip_metadata = self._get_zip_metadata()
+
+    def _initialize_s3_client(self):
+        import botocore.session
+        session = botocore.session.get_session()
+        return session.create_client(
+            "s3",
+            aws_access_key_id=self.AWS_ACCESS_KEY_ID,
+            aws_secret_access_key=self.AWS_SECRET_ACCESS_KEY,
+            aws_session_token=self.AWS_SESSION_TOKEN,
+            endpoint_url=self.AWS_ENDPOINT_URL,
+            region_name=self.AWS_REGION_NAME,
+            use_ssl=self.AWS_USE_SSL,
+            verify=self.AWS_VERIFY,
+        )
+
+    def get_object(self, range_bytes: str) -> bytes:
+        response = self.s3_client.get_object(Bucket=self.bucket, Key=self.path, Range=range_bytes)
+        return response["Body"].read()
+
+    def get_file_info(self) -> dict:
+        return self.s3_client.head_object(Bucket=self.bucket, Key=self.path)
+
+    def fetch_file_part(self, file_name: str, offset: int, size: int) -> bytes:
+        metadata = self.zip_metadata[file_name]
+        range_bytes = self._calculate_range(metadata, offset, size)
+        data = self.get_object(range_bytes)
+
+        return gzip.open(BytesIO(data)).read() if file_name.endswith(".gz") else data
+
+    def fetch_file(self, file_name: str) -> bytes:
+        metadata = self.zip_metadata[file_name]
+        range_bytes = self._calculate_range(metadata, 0, metadata["compressed_size"])
+        return self.get_object(range_bytes)
+
+    def _calculate_range(self, metadata: dict, offset: int, size: int) -> str:
+        start = metadata["header_offset"] + metadata["file_header_length"] + offset
+        end = start + size - 1
+        return f"bytes={start}-{end}"
+
+    @cached_property
+    def zip_exists(self) -> bool:
+        return bool(self.get_file_info())
+
+
+class ZipStorageHandlerFactory:
+    @staticmethod
+    def get_handler(uri: str, settings) -> ZipStorageHandler:
+        if uri.startswith('s3://'):
+            handler = S3ZipStorageHandler
+            handler.AWS_ACCESS_KEY_ID = settings["AWS_ACCESS_KEY_ID"]
+            handler.AWS_SECRET_ACCESS_KEY = settings["AWS_SECRET_ACCESS_KEY"]
+            handler.AWS_SESSION_TOKEN = settings["AWS_SESSION_TOKEN"]
+            handler.AWS_ENDPOINT_URL = settings["AWS_ENDPOINT_URL"]
+            handler.AWS_REGION_NAME = settings["AWS_REGION_NAME"]
+            handler.AWS_USE_SSL = settings["AWS_USE_SSL"]
+            handler.AWS_VERIFY = settings["AWS_VERIFY"]
```

Comment:
There is also a setting related to whether to build the final S3 endpoint URL with the bucket name in it or not (path style), that is quite important to access non-AWS S3 endpoints.

Reply:
Is this an existing setting in Scrapy? As far as I can see all the existing settings are passed on when creating the s3 client. From your comment I would guess you are referring to `AWS_ENDPOINT_URL` which we also do implement. Am I missing something here?

Reply:
Ah, I see we don't need this setting - apparently the default path style is fine. Nothing to do here then!